### DISCUSSION TITLE
Possibility to throw Directus Exceptions in (Event Filter) Flows

### DISCUSSION THREAD

**Discussion by timspi** [directus/directus#16342] [Feature Requests]
When using an Event triggered flow as a filter, I think at the moment, it is only possible to cancel the event by ending on a reject path. This could be a false condition or a thrown error in Run Script. However, as these errors are never an instance of the `BaseException` class, the error returned by the API will always be 500 Internal Server Error without any further information.

It would be very useful if there is a possibility to specify an error message in the flow, e.g. with a new operation "Throw exception" where the status code and error message could be configured. For example, when doing a check with a Condition operation, instead of just ending the flow on the reject path, a "Throw exception" could be added and the error message could be changed, e.g. to 400 Bad request with a custom error message.

Another option is to somehow import `BaseException` (or any of the other preconfigured exceptions) inside of Run Script to be able to throw them and have the error message displayed. I am not sure though if this is possible with NodeVM. I tried using the env variable `FLOWS_EXEC_ALLOWED_MODULES` and import `BaseExcpetion` from `@directus/shared`, however it errors with VMError: Cannot find module '@directus/shared'. I am not sure where it is looking for the module. But this way it probably would not work anyways as the instanceof method would probably return false when the module is imported in different ways. Another option might be to pass the exceptions from the context property to the VM so that they could be used there. I am not sure though if that would solve the instanceof problem. Maybe the cleaner way is to just use the previously proposed "Throw exception" operation on the reject path of the Run Script operation.

**Comment by timspi** [comment]
I have just tested the option to pass the exceptions property of the context to the VM and it works fine.

I changed the file _api/src/operations/exec/index.ts_ to include the exceptions in the data payload:
<img src="https://user-images.githubusercontent.com/24900840/200309942-a93426af-cb7d-430a-911b-6fcc073c9a0f.png" alt="image" width="500"/>

Then, inside a flow, the exception can be used in Run Script:
```
module.exports = async function(data) {
    const { BaseException } = data.$exceptions;
    throw new BaseException('Custom error message', 400, 'BAD_REQUEST');
    
    return {};
}
```

Like that, _api/src/middleware/error-handler.ts_ returns the error correctly:
![image](https://user-images.githubusercontent.com/24900840/200311303-ea6955e6-0431-4ebf-bcf2-5d41cf10833e.png)

**Reply by timspi** [reply]
I found a better way to pass the exceptions into the VM. There is the `freeze` function with which a constant global can be defined in the VM. So by adding `vm.freeze(exceptions, '$exceptions');` after creating the `NodeVM` instance, the following code could be used in Run Script to throw a custom error message:
```
module.exports = async function(data) {
    const { BaseException } = $exceptions;
    throw new BaseException('Custom error message', 400, 'BAD_REQUEST');
    
    return {};
}
```

Or a function could be passed, e.g. like so:
```
vm.freeze((...params) => {
    throw new BaseException(...params);
}, '$throwCustomError');
```
Then, inside Run Script:
```
module.exports = async function(data) {
    $throwCustomError('Custom error message', 400, 'BAD_REQUEST');
    
    return {};
}
```

**Comment by rijkvanzanten** [comment]
I think we should do this as another operation, so you can format the error as per your requirements through options 🤔

**Comment by shipsides-sparknel** [comment]
I am out of my area of expertise so please feel free to school me:
Inside a run script operation:
`throw new Error('My custom error message.');`
when running as admin:
![image](https://user-images.githubusercontent.com/106218118/202969902-e00cff1f-6a98-4a91-8992-5b0bd92406cd.png)
when running as a non-admin user:
![image](https://user-images.githubusercontent.com/106218118/202969937-cf3b527f-7ef4-45c4-ade8-baf87e479c0a.png)

while the ability to display a non-500 message is useful, displaying a detailed custom message with meaning to the user satisfies the majority of requirements. why does this function work as an admin but not for non-admins? is there a secret permission or setting that would enable the full message to be shown to non-admins?

**Reply by timspi** [reply]
This is a security mechanism so that a potential stack trace, which might include sensitive information, is not leaked to non-admin users. To include the message for all users, `BaseException` has to be used instead of `Error`. However, this Directus internal class is not available inside Run Script at the moment, therefore it is not yet possible to achieve this. Hence I started this discussion.

The relevant section in the codebase is in file `api/src/middleware/error-handler.ts`:
```
if (req.accountability?.admin === true) {
    payload = {
        errors: [
            {
                message: err.message,
                extensions: {
                    code: 'INTERNAL_SERVER_ERROR',
                    ...err.extensions,
                },
            },
        ],
    };
} else {
    payload = {
        errors: [
            {
                message: 'An unexpected error occurred.',
                extensions: {
                    code: 'INTERNAL_SERVER_ERROR',
                },
            },
        ],
    };
}
```

But I think this behaviour is fine and should not be changed. The way to add custom error messages is to pass the `BaseException` class into Run Script and/or add the new proposed "Throw exception" operation.

**Reply by shipsides-sparknel** [reply]
thank you for explaining this 👍

**Comment by 53RG1005** [comment]
How to create custom error message in Flow? Maybe someone knows?

throw new Error('Message') returns only error in panel, but via api still INTERNAL_SERVER_ERROR

**Comment by grefrit** [comment]
It seems that it is not possible anymore, because of recent changes is `flows.ts`
https://github.com/directus/directus/blob/cdf8d94ec1e9d6e948660c890e3e7d1876fc6890/api/src/flows.ts#L426-L427

This piece of code converts `BaseException` to just plain object, so this `data` is not instance of `BaseException`. 

So `error-handler` cannot handle that here
https://github.com/directus/directus/blob/cdf8d94ec1e9d6e948660c890e3e7d1876fc6890/api/src/middleware/error-handler.ts#L19-L20

and here
https://github.com/directus/directus/blob/cdf8d94ec1e9d6e948660c890e3e7d1876fc6890/api/src/middleware/error-handler.ts#L43-L58

**Reply by rijkvanzanten** [reply]
~~Mind opening a new issue? ty! (cc @nickrum)~~ You already did, thanks! 😄  👉🏻 https://github.com/directus/directus/issues/18338

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
